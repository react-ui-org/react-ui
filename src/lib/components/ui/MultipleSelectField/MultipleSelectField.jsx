import PropTypes from 'prop-types';
import React from 'react';
import transferProps from '../../../utils/transferProps';
import withForwardedRef from '../withForwardedRef';
import styles from './MultipleSelectField.scss';

export const MultipleSelectField = (props) => {
  let labelVisibilityClass = '';
  let rootFullWidthClass = '';
  let rootLayoutClass = '';
  let rootRequiredClass = '';
  let rootSizeClass = '';
  let rootValidationStateClass = '';
  let rootVariantClass = '';

  if (!props.isLabelVisible) {
    labelVisibilityClass = styles.isLabelHidden;
  }

  if (props.fullWidth) {
    rootFullWidthClass = styles.isRootFullWidth;
  }

  if (props.layout === 'horizontal') {
    rootLayoutClass = styles.rootLayoutHorizontal;
  } else if (props.layout === 'vertical') {
    rootLayoutClass = styles.rootLayoutVertical;
  }

  if (props.required) {
    rootRequiredClass = styles.isRootRequired;
  }

  if (props.size === 'small') {
    rootSizeClass = styles.rootSizeSmall;
  } else if (props.size === 'medium') {
    rootSizeClass = styles.rootSizeMedium;
  } else if (props.size === 'large') {
    rootSizeClass = styles.rootSizeLarge;
  }

  if (props.validationState === 'invalid') {
    rootValidationStateClass = styles.isRootStateInvalid;
  } else if (props.validationState === 'valid') {
    rootValidationStateClass = styles.isRootStateValid;
  } else if (props.validationState === 'warning') {
    rootValidationStateClass = styles.isRootStateWarning;
  }

  if (props.variant === 'filled') {
    rootVariantClass = styles.rootVariantFilled;
  } else if (props.variant === 'outline') {
    rootVariantClass = styles.rootVariantOutline;
  }

  const propsToTransfer = transferProps(
    props,
    ['changeHandler', 'disabled', 'fullWidth', 'helperText', 'id', 'isLabelVisible',
      'label', 'layout', 'options', 'required', 'size', 'validationState', 'value', 'variant'],
  );

  return (
    <label
      className={(`
        ${styles.root}
        ${rootFullWidthClass}
        ${rootLayoutClass}
        ${rootRequiredClass}
        ${rootSizeClass}
        ${rootValidationStateClass}
        ${rootVariantClass}
      `).trim()}
      htmlFor={props.id}
      id={`${props.id}__label`}
    >
      <div
        className={(`
          ${styles.label}
          ${labelVisibilityClass}
        `).trim()}
        id={`${props.id}__labelText`}
      >
        {props.label}
      </div>
      <div className={styles.inputContainer}>
        <select
          {...propsToTransfer}
          className={styles.input}
          disabled={props.disabled}
          id={props.id}
          multiple
          onChange={props.changeHandler}
          ref={props.forwardedRef}
          required={props.required}
          value={props.value}
        >
          {
            props.options.map((option) => (
              <option
                disabled={option.disabled}
                id={`${props.id}__item__${option.value}`}
                key={option.value}
                value={option.value}
              >
                {option.label}
              </option>
            ))
          }
        </select>
        {props.variant === 'filled' && (
          <div className={styles.bottomLine} />
        )}
      </div>
      {props.helperText && (
        <div
          className={styles.helperText}
          id={`${props.id}__helperText`}
        >
          {props.helperText}
        </div>
      )}
    </label>
  );
};

MultipleSelectField.defaultProps = {
  changeHandler: null,
  disabled: false,
  forwardedRef: undefined,
  fullWidth: false,
  helperText: null,
  isLabelVisible: true,
  layout: 'vertical',
  required: false,
  size: 'medium',
  validationState: null,
  value: [],
  variant: 'outline',
};

MultipleSelectField.propTypes = {
  changeHandler: PropTypes.func,
  disabled: PropTypes.bool,
  forwardedRef: PropTypes.func,
  fullWidth: PropTypes.bool,
  helperText: PropTypes.string,
  id: PropTypes.string.isRequired,
  isLabelVisible: PropTypes.bool,
  label: PropTypes.string.isRequired,
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  options: PropTypes.arrayOf(PropTypes.shape({
    disabled: PropTypes.bool,
    label: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
  })).isRequired,
  required: PropTypes.bool,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  value: PropTypes.arrayOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ])),
  variant: PropTypes.oneOf(['filled', 'outline']),
};

export default withForwardedRef(MultipleSelectField);
