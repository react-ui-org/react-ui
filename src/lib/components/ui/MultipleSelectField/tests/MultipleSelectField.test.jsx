import React from 'react';
import {
  shallow,
  mount,
} from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import { MultipleSelectField } from '../MultipleSelectField';

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(<MultipleSelectField
      id="test"
      label="label"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
        {
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly only one option', () => {
    const tree = shallow(<MultipleSelectField
      id="test"
      label="label"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
      ]}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(<MultipleSelectField
      id="test"
      inFormLayout
      isLabelVisible={false}
      label="label"
      layout="horizontal"
      validationState="invalid"
      value={['ch1']}
      helperText="some help"
      variant="filled"
      size="large"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
        {
          disabled: true,
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
      fullWidth
      disabled
      required
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with one selected option', () => {
    const tree = shallow(<MultipleSelectField
      id="test"
      label="label"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
        {
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
      value={['ch1']}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with two selected options', () => {
    const tree = shallow(<MultipleSelectField
      id="test"
      label="label"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
        {
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
      value={['ch1', 'ch2']}
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls changeHandler() on changing selected option', () => {
    const spy = sinon.spy();
    const component = mount(<MultipleSelectField
      label="label"
      id="id"
      options={[
        {
          label: 'choice 1',
          value: 'ch1',
        },
        {
          label: 'choice 2',
          value: 'ch2',
        },
      ]}
      changeHandler={spy}
      value={['ch1']}
    />);

    component
      .find('select')
      .simulate('change', {
        preventDefault: () => {},
        target: { value: 'ch2' },
      });
    expect(spy.calledOnce).toEqual(true);
  });
});
