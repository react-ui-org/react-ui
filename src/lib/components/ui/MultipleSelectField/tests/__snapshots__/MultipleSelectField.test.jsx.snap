// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<label
  className="root
        
        
        rootLayoutVertical
        
        rootSizeMedium
        
        rootVariantOutline"
  htmlFor="test"
  id="test__label"
>
  <div
    className="label"
    id="test__labelText"
  >
    label
  </div>
  <div
    className="inputContainer"
  >
    <select
      className="input"
      disabled={false}
      id="test"
      multiple={true}
      onChange={null}
      required={false}
      value={Array []}
    >
      <option
        id="test__item__ch1"
        key="ch1"
        value="ch1"
      >
        choice 1
      </option>
      <option
        id="test__item__ch2"
        key="ch2"
        value="ch2"
      >
        choice 2
      </option>
    </select>
  </div>
</label>
`;

exports[`rendering renders correctly only one option 1`] = `
<label
  className="root
        
        
        rootLayoutVertical
        
        rootSizeMedium
        
        rootVariantOutline"
  htmlFor="test"
  id="test__label"
>
  <div
    className="label"
    id="test__labelText"
  >
    label
  </div>
  <div
    className="inputContainer"
  >
    <select
      className="input"
      disabled={false}
      id="test"
      multiple={true}
      onChange={null}
      required={false}
      value={Array []}
    >
      <option
        id="test__item__ch1"
        key="ch1"
        value="ch1"
      >
        choice 1
      </option>
    </select>
  </div>
</label>
`;

exports[`rendering renders correctly with all props 1`] = `
<label
  className="root
        isRootFullWidth
        isRootInFormLayout
        rootLayoutHorizontal
        isRootRequired
        rootSizeLarge
        isRootStateInvalid
        rootVariantFilled"
  htmlFor="test"
  id="test__label"
>
  <div
    className="label
          isLabelHidden"
    id="test__labelText"
  >
    label
  </div>
  <div
    className="inputContainer"
  >
    <select
      className="input"
      disabled={true}
      id="test"
      multiple={true}
      onChange={null}
      required={true}
      value={
        Array [
          "ch1",
        ]
      }
    >
      <option
        id="test__item__ch1"
        key="ch1"
        value="ch1"
      >
        choice 1
      </option>
      <option
        disabled={true}
        id="test__item__ch2"
        key="ch2"
        value="ch2"
      >
        choice 2
      </option>
    </select>
    <div
      className="bottomLine"
    />
  </div>
  <div
    className="helperText"
    id="test__helperText"
  >
    some help
  </div>
</label>
`;

exports[`rendering renders correctly with one selected option 1`] = `
<label
  className="root
        
        
        rootLayoutVertical
        
        rootSizeMedium
        
        rootVariantOutline"
  htmlFor="test"
  id="test__label"
>
  <div
    className="label"
    id="test__labelText"
  >
    label
  </div>
  <div
    className="inputContainer"
  >
    <select
      className="input"
      disabled={false}
      id="test"
      multiple={true}
      onChange={null}
      required={false}
      value={
        Array [
          "ch1",
        ]
      }
    >
      <option
        id="test__item__ch1"
        key="ch1"
        value="ch1"
      >
        choice 1
      </option>
      <option
        id="test__item__ch2"
        key="ch2"
        value="ch2"
      >
        choice 2
      </option>
    </select>
  </div>
</label>
`;

exports[`rendering renders correctly with two selected options 1`] = `
<label
  className="root
        
        
        rootLayoutVertical
        
        rootSizeMedium
        
        rootVariantOutline"
  htmlFor="test"
  id="test__label"
>
  <div
    className="label"
    id="test__labelText"
  >
    label
  </div>
  <div
    className="inputContainer"
  >
    <select
      className="input"
      disabled={false}
      id="test"
      multiple={true}
      onChange={null}
      required={false}
      value={
        Array [
          "ch1",
          "ch2",
        ]
      }
    >
      <option
        id="test__item__ch1"
        key="ch1"
        value="ch1"
      >
        choice 1
      </option>
      <option
        id="test__item__ch2"
        key="ch2"
        value="ch2"
      >
        choice 2
      </option>
    </select>
  </div>
</label>
`;
